### ISSUE TITLE
multiple expression error doesn't show the multiple expressions

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#806]
I'd have expected "block contains multiple expressions" to be a showcase of multi-colored highlighting, since it would show what each of the expressions is. But instead I get this output:

![image](https://cloud.githubusercontent.com/assets/75030/17277792/dba4cb9c-5719-11e6-84b7-05bf6ed03295.png)

If I didn't already see what the multiple expressions were, this huge block wouldn't help me do so. Indeed, it would be tempting to just wrap a `block` since it's much more expedient.

**Comment by blerner** [comment]
First, I think showing all the individual expressions will be a fruit salad nightmare of colors-- are any of them important? Which do I look at? Why should I care? What do I fix?  Additionally, this error needs to work for trickier situations like cases expressions where some but not all branches have multiple expressions in them -- coloring yet more expressions will be even more distracting and unhelpful. If the programmer decides to mark the whole thing as a block, that's not inherently a bad decision. The biggest readability problem with this message is just showing too much code as a monolithic block; making it more colorful won't alleviate that.

**Comment by shriram** [comment]
Yes, showing every single thing would indeed be a problem — maybe. But that is not an argument for the current design. If I already know what the problem is by inspection, great. But I may not understand the grammar of the language fully. Students are routinely confused by the status of the "let bindings" in blocks: are those statements, what? The parser knows; it is ground truth; furthermore, its internal state is the only thing that really matters. By not revealing it, we're withholding information that is actually useful. At the very least we should highlight the first two blocks and say "and then there are more" (and maybe even stop highlighting the rest, to avoid suggesting everything else is one block, or perhaps even put it in a show more if that's possible).

Bottom line: yes, showing every block up front may be bad. But showing just one color and thus hiding the parser state is not helpful to novices who don't understand the language grammar and hence won't even know where to start to try to fix things.

**Comment by shriram** [comment]
Note that I'm not talking about the case where the fix is to just mark everything as one `block`. (If I have two `ask` expressions, I don't want to just blindly ignore one of them.) It's the case where the user _intended_ for one expression and can't figure out why Pyret thinks there are more than one.

**Comment by jswrenn** [comment]
I think the wrong things are being highlighted here. The error message complains about "multiple expressions", but not all of the source locations this error is getting passed correspond to expressions. The only things that should be highlighted are the expressions on lines 24 and 37.

That would help, but doesn't address the issue of distinguishing the expressions with multiple colors. If the two expressions are on adjacent lines, it won't be possible to tell the difference between one highlight that's two lines long, or two one-line-long highlights. The trouble is figuring out how to introduce the colors. I can only give the `multiple expressions` anchor tag one color.

**Comment by blerner** [comment]
In the interests of not getting bogged down by technical details, I've committed a change (1ad09024accbcd08c03a063318fbf288bb197c1f) that gives you access to the blocks themselves, instead of their locations.  Go crazy on them, filtering out the expressions you want to blame, or whatever.

**Comment by jswrenn** [comment]
@blerner I don't think this helps. Notice how in Shriram's screenshot, the highlights extend all the way to the gutter after that first highlighted line? This is because there is only one thing being highlighted. If you were to type in one of those empty lines, what you typed would be highlighted, too, since it's within the single, large highlighted span.

**Comment by blerner** [comment]
You're missing the point: instead of giving you the block's locations, I'm giving you the ast of the blocks.  That means you have their statements, individually, and their locations, individually.  How you choose to give them colors and associate the colors with text in the error wording is up to you.  But you now have _all_ the possible information you could possibly need, for this message.
